### ISSUE TITLE
Feature Request: Ability to handle tag before adding

### ISSUE THREAD

**Issue by shyamal890** [ng-select/ng-select#309]
https://github.com/ng-select/ng-select/issues/220 only addresses 'NEW' tags. This should be extended for any options in dropdown. 

`onOptionSelection` - Expression to evaluate that will be invoked before adding an `option` from the dropdown to `ngModel`. The `option` model is available as `$option`. This method must return either a boolean value or a promise. If either a false value or a rejected promise is returned, the `option` will not be added.

In conjunction with `onOptionSelection`, `onOptionSelected` can also be added.

**Example Usecase:**

- Say you have more than 20 `options` and you want to add all of them together in `ngModel`. You can have an `option` saying `Add All` which adds all the available options to the `ngModel` at once. This is only possible is we can handle `Add All` `option` object before being adding it to `ngModel`.
- `Options` like `Add All`, `None`, etc would only be visible when the user types matching/containing phrase.
- Adding some `options` like `critical`, `important` should only be possible if the data meets certain criteria.

**Comment by shyamal890** [comment]
@anjmao Can you please add this feature.

**Comment by anjmao** [comment]
@shyamal890 Hi, in you use case, maybe you can use header or footer template and add button `Add all`, which sounds like a normal solution to me?

**Comment by shyamal890** [comment]
Elaborating on the usecase: 

1. Options like `Add All`, `None`, etc should only be visible when the user types matching/containing phrase.
2. Adding an option like `critical` should only be possible if the data meets certain criteria. For this we need to make a request to the server for checking possibility.

**Comment by shyamal890** [comment]
@anjmao thoughts on when this would be available?

**Comment by anjmao** [comment]
@shyamal890 This is very specific use case. You could have optiions like `Add All` and clicking on them you can programatically push all options to ngModel.

**Comment by shyamal890** [comment]
I personally think this would be required by many projects and this is reflected in the following libraries too:

[Bootstrap Tags Input](https://bootstrap-tagsinput.github.io/bootstrap-tagsinput/examples/) too has `beforeItemAdd` that provides a way to handle option before adding to model
[React Select](https://github.com/JedWatson/react-select) has `handleChange` 
[ngTagsInput](http://mbenford.github.io/ngTagsInput/documentation/api) has `onTagAdding` to support this usecase.

**Comment by varnastadeus** [comment]
Firstly let's make clear with naming:
'Tag'  only exists while it's not added, when you add it, it is regular 'option'  and has no differences among other items.
Can you rephrase this feature request having this in mind?

**Comment by shyamal890** [comment]
@varnastadeus I updated the feature request given the above mentioned semantics.

**Comment by shyamal890** [comment]
Wonder when this feature would be available, currently implementing a workaround with footer which is just a hack. Can't select footer on `Enter`.
